### PR TITLE
EPMLABSBRN-84-API-spec

### PR DESCRIPTION
**Pull request by H1D**
[EPMLABSBRN-84](https://jira.epam.com/jira/browse/EPMLABSBRN-84)

**Description**:
Ads API specification according in [RAML](https://raml.org) format

Spec rendered to HTML – [api.html.zip](https://github.com/Brain-up/brn/files/3780865/api.html.zip)

### CODE REVIEW

File: api-contract/samples/tasks_by_exercise.json [hunk 1]
```diff
@@ -0,0 +1,229 @@
+{
+ "data": [
+  {
```
Comment (goopnigoop) [inline api-contract/samples/tasks_by_exercise.json]: I'd like to suggest this structure of task:
    {
        "id": 1,
        "name": "task",
        "correctAnswer": {
            "id": 1,
            "audioFileUrl": "noise_0db/бум.mp3",
            "word": "бум",
            "pictureFileUrl": "",
            "soundsCount": 1
        },
        "serialNumber": 2,
        "answerOptions": []
    },
    {
        "id": 2,
        "name": "task",
        "correctAnswer": {
            "id": 2,
            "audioFileUrl": "noise_0db/бал.mp3",
            "word": "бал",
            "pictureFileUrl": "",
            "soundsCount": 1
        },
        "serialNumber": 1,
        "answerOptions": []
    },
    {
        "id": 3,
        "name": "task",
        "correctAnswer": {
            "id": 3,
            "audioFileUrl": "no_noise/бал.mp3",
            "word": "бал",
            "pictureFileUrl": "",
            "soundsCount": 1
        },
        "serialNumber": 1,
        "answerOptions": [
            {
                "id": 8,
                "audioFileUrl": "no_noise/бум.mp3",
                "word": "бум",
                "pictureFileUrl": "",
                "soundsCount": 1
            },
            {
                "id": 9,
                "audioFileUrl": "noise_0db/вить.mp3",
                "word": "вить",
                "pictureFileUrl": "",
                "soundsCount": 1
            },
            {
                "id": 4,
                "audioFileUrl": "no_noise/быль.mp3",
                "word": "быль",
                "pictureFileUrl": "",
                "soundsCount": 1
            },
            {
                "id": 10,
                "audioFileUrl": "",
                "word": "сад",
                "pictureFileUrl": "",
                "soundsCount": 1
            },
            {
                "id": 11,
                "audioFileUrl": "noise_0db/быль.mp3",
                "word": "быль",
                "pictureFileUrl": "",
                "soundsCount": 1
            },
            {
                "id": 5,
                "audioFileUrl": "noise_0db/гад.mp3",
                "word": "гад",
                "pictureFileUrl": "",
                "soundsCount": 1
            },
            {
                "id": 1,
                "audioFileUrl": "noise_0db/бум.mp3",
                "word": "бум",
                "pictureFileUrl": "",
                "soundsCount": 1
            },
            {
                "id": 6,
                "audioFileUrl": "no_noise/гад.mp3",
                "word": "гад",
                "pictureFileUrl": "",
                "soundsCount": 1
            },
            {
                "id": 7,
                "audioFileUrl": "no_noise/вить.mp3",
                "word": "вить",
                "pictureFileUrl": "",
                "soundsCount": 1
            }
        ]
    }

Have we decided to use this structure or we'll add one more wrapper to process it to structure described in json?
@ElenaSpb

File: api-contract/api.raml [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+      seriesId:
+        description: "ID of the series exercises belongs to"
+        required: true
+        type: string
+        examle: "42"
+    responses:
+      200:
+        body:
+          example: !include samples/exercises_by_series.json
+
```
Comment (goopnigoop) [inline api-contract/api.raml]: I think it's just a beginning and we will add for example "/{exerciseId}:" in the future? The same question to get a task by id.

Comment (rpcep) [reply api-contract/api.raml]: I assumed that was the case, some endpoints that are already implemented are missing, but this is a good beginning.

Comment (H1D) [reply api-contract/api.raml]: I have listed ones we really need for know.

File: api-contract/samples/tasks_by_exercise.json (hunk 1 above)
Comment (H1D) [reply api-contract/samples/tasks_by_exercise.json]: @goopnigoop JSON you provided is seems broken...

Also could you please add indentation using this for example –https://jsonformatter.curiousconcept.com/ 

and use this button to insert code:
image

File: api-contract/samples/exercises_by_series.json [hunk 3]
```diff
@@ -0,0 +1,40 @@
+{
+ "data": [
+  {
+   "id": "1",
+   "name": "однослоговые слова",
+   "series": "1",
+   "tasks": [
```
Comment (ElenaSpb) [inline api-contract/samples/exercises_by_series.json]: @H1D does you mean include tasks by default?

File: api-contract/samples/groups.json [hunk 4]
```diff
@@ -0,0 +1,46 @@
+{
+ "data": [
+  {
+   "name": "Group 1",
+   "id": "1",
+   "series": [
```
Comment (ElenaSpb) [inline api-contract/samples/groups.json]: so series list by default too
and as in all others entities

File: api-contract/samples/tasks_by_exercise.json (hunk 1 above)
Comment (goopnigoop) [reply api-contract/samples/tasks_by_exercise.json]:
```
[
  {
    "id": 1,
    "name": "task",
    "correctAnswer": {
      "id": 1,
      "audioFileUrl": "noise_0db/бум.mp3",
      "word": "бум",
      "pictureFileUrl": "",
      "soundsCount": 1
    },
    "serialNumber": 2,
    "answerOptions": []
  },
  {
    "id": 2,
    "name": "task",
    "correctAnswer": {
      "id": 2,
      "audioFileUrl": "noise_0db/бал.mp3",
      "word": "бал",
      "pictureFileUrl": "",
      "soundsCount": 1
    },
    "serialNumber": 1,
    "answerOptions": []
  },
  {
    "id": 3,
    "name": "task",
    "correctAnswer": {
      "id": 3,
      "audioFileUrl": "no_noise/бал.mp3",
      "word": "бал",
      "pictureFileUrl": "",
      "soundsCount": 1
    },
    "serialNumber": 1,
    "answerOptions": [
      {
        "id": 8,
        "audioFileUrl": "no_noise/бум.mp3",
        "word": "бум",
        "pictureFileUrl": "",
        "soundsCount": 1
      },
      {
        "id": 9,
        "audioFileUrl": "noise_0db/вить.mp3",
        "word": "вить",
        "pictureFileUrl": "",
        "soundsCount": 1
      },
      {
        "id": 4,
        "audioFileUrl": "no_noise/быль.mp3",
        "word": "быль",
        "pictureFileUrl": "",
        "soundsCount": 1
      },
      {
        "id": 10,
        "audioFileUrl": "",
        "word": "сад",
        "pictureFileUrl": "",
        "soundsCount": 1
      },
      {
        "id": 11,
        "audioFileUrl": "noise_0db/быль.mp3",
        "word": "быль",
        "pictureFileUrl": "",
        "soundsCount": 1
      },
      {
        "id": 5,
        "audioFileUrl": "noise_0db/гад.mp3",
        "word": "гад",
        "pictureFileUrl": "",
        "soundsCount": 1
      },
      {
        "id": 1,
        "audioFileUrl": "noise_0db/бум.mp3",
        "word": "бум",
        "pictureFileUrl": "",
        "soundsCount": 1
      },
      {
        "id": 6,
        "audioFileUrl": "no_noise/гад.mp3",
        "word": "гад",
        "pictureFileUrl": "",
        "soundsCount": 1
      },
      {
        "id": 7,
        "audioFileUrl": "no_noise/вить.mp3",
        "word": "вить",
        "pictureFileUrl": "",
        "soundsCount": 1
      }
    ]
  },
  {
    "id": 4,
    "name": "task",
    "correctAnswer": {
      "id": 8,
      "audioFileUrl": "no_noise/бум.mp3",
      "word": "бум",
      "pictureFileUrl": "",
      "soundsCount": 1
    },
    "serialNumber": 2,
    "answerOptions": [
      {
        "id": 3,
        "audioFileUrl": "no_noise/бал.mp3",
        "word": "бал",
        "pictureFileUrl": "",
        "soundsCount": 1
      },
      {
        "id": 9,
        "audioFileUrl": "noise_0db/вить.mp3",
        "word": "вить",
        "pictureFileUrl": "",
        "soundsCount": 1
      },
      {
        "id": 4,
        "audioFileUrl": "no_noise/быль.mp3",
        "word": "быль",
        "pictureFileUrl": "",
        "soundsCount": 1
      },
      {
        "id": 10,
        "audioFileUrl": "",
        "word": "сад",
        "pictureFileUrl": "",
        "soundsCount": 1
      },
      {
        "id": 11,
        "audioFileUrl": "noise_0db/быль.mp3",
        "word": "быль",
        "pictureFileUrl": "",
        "soundsCount": 1
      },
      {
        "id": 5,
        "audioFileUrl": "noise_0db/гад.mp3",
        "word": "гад",
        "pictureFileUrl": "",
        "soundsCount": 1
      },
      {
        "id": 2,
        "audioFileUrl": "noise_0db/бал.mp3",
        "word": "бал",
        "pictureFileUrl": "",
        "soundsCount": 1
      },
      {
        "id": 6,
        "audioFileUrl": "no_noise/гад.mp3",
        "word": "гад",
        "pictureFileUrl": "",
        "soundsCount": 1
      },
      {
        "id": 7,
        "audioFileUrl": "no_noise/вить.mp3",
        "word": "вить",
        "pictureFileUrl": "",
        "soundsCount": 1
      }
    ]
  }
]
```

File: api-contract/samples/exercises_by_series.json (hunk 3 above)
Comment (H1D) [reply api-contract/samples/exercises_by_series.json]: Not exactly. `include` means including objects itself.

Tasks `ids` meats to be always listed when we fetch exercise

**Pull request merged**